### PR TITLE
[BEAM-7366] Determine Spotless target based on project's SourceSet

### PR DESCRIPTION
**Pull request by mxm**
Spotless currently traverses the project files to discover the source
files. This can causes problems when the project directory uses a source set
which spans outside the project folder. This is the case for the Flink Runner
projects which pull in their main source from `runners/flink` with the version
specific code being in `runners/flink/1.x`.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by mxm** [comment]
Test failures due https://issues.apache.org/jira/browse/BEAM-7367, fix pending #8625.

**Comment by mxm** [comment]
Run JavaPortabilityApi PreCommit

**Comment by mxm** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by mxm** [comment]
Run JavaPortabilityApi PreCommit

**Comment by mxm** [comment]
Run Python_PVR_Flink PreCommit

**Comment by mxm** [comment]
Run Python_PVR_Flink PreCommit

**Comment by mxm** [comment]
@adude3141 Let me know what you think about this when you get a chance.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -818,13 +818,7 @@ class BeamModulePlugin implements Plugin<Project> {
...
           licenseHeader javaLicenseHeader
           googleJavaFormat('1.7')
-          target project.fileTree(project.projectDir) {
-            include '**/*.java'
-            exclude '**/archetype-resources/src/**'
-            exclude '**/build/generated/**'
-            exclude '**/build/generated-src/**'
-            exclude '**/build/generated-*-avro-*/**'
-          }
+          target project.sourceSets.main.allJava, project.sourceSets.test.allJava
```
Comment (mxm) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Just to highlight the key change here; the rest is spotless re-formatting.

**Comment by adude3141** [comment]
Diff between both filesets:

```diff
2106d2105
< /Users/michel/GitHub/adude3141/beam/sdks/java/core/src/test/avro/org/apache/beam/sdk/state/StateContextsTest.java
3444,3445d3442
< /Users/michel/GitHub/adude3141/beam/sdks/java/maven-archetypes/starter/build/resources/test/projects/basic/reference/src/main/java/it/pkg/StarterPipeline.java
< /Users/michel/GitHub/adude3141/beam/sdks/java/maven-archetypes/starter/src/test/resources/projects/basic/reference/src/main/java/it/pkg/StarterPipeline.java
```
I.e. this three files will not be under spotless anymore. Need to understand first, whether this imposes a problem.

**Comment by adude3141** [comment]
Could it be, that the file
`beam/sdks/java/core/src/test/avro/org/apache/beam/sdk/state/StateContextsTest.java` is just misplaced and should be moved to to `beam/sdks/java/core/src/test/java/org/apache/beam/sdk/state/StateContextsTest.java`? I can not find any reason, why this should be placed there. Especially as this is *not* the place where Gradle is looking for tests.

**Comment by mxm** [comment]
Thanks for creating a diff for the old and the new target files. `StateContextsTest` is not correctly placed and will indeed not be executed as part of the tests. It's missing the `main` directory.

**Comment by mxm** [comment]
I tried running the test via ```
 gradle :sdks:java:core:test --tests "**StateContextsTest**"``` and it would not execute. Will move the test to the main test directory.

**Comment by mxm** [comment]
I think this looks good now. What do you think @adude3141?

Test failure for Flink PVR has a fix: #8633.

**Comment by mxm** [comment]
Run Python_PVR_Flink PreCommit

**Comment by adude3141** [comment]
Sorry for the delay.

Unfortunately I am a bit confused currently. Executing
```
./gradlew clean build -p sdks/java/core/
```
fails on my machine now with spotless complaining about the antlr generated files. Which I honestly do not understand, as those files did not show up during diffing.

So we might need to keep the exclusion filter anyway and change the implementation to something like
```
          def union = project.files([]).asFileTree
          project.sourceSets.each { sourceset -> union += sourceset.allJava }
          target union.matching {
            include '**/*.java'
            exclude '**/archetype-resources/src/**'
            exclude '**/build/generated/**'
            exclude '**/build/resources/**'
            exclude '**/build/generated-src/**'
            exclude '**/build/generated-*-avro-*/**'
          }.files
```

**Comment by mxm** [comment]
Thanks @adude3141 for the suggestion. Sounds good to me. Not sure why the generated files didn't show up in the source set, but it doesn't hurt to include both the project files and the source sets and run the filter on these.

It didn't need the resource filter to let Spotless pass. We probably want to run spotless on Java resources that are included in the archetypes.

**Comment by adude3141** [comment]
Run Python_PVR_Flink PreCommit

**Comment by adude3141** [comment]
Run Python PreCommit

**Comment by mxm** [comment]
Run CommunityMetrics PreCommit

**Comment by mxm** [comment]
Community metrics are broken (unrelated): https://builds.apache.org/job/beam_PreCommit_CommunityMetrics_Phrase/3/

**Pull request merged**